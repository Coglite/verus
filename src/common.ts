import * as kindOf from 'kind-of'
import { Errors, errorMessage } from './errors'

export type Kind =
    'undefined'
    | 'null'
    | 'boolean'
    | 'buffer'
    | 'number'
    | 'string'
    | 'arguments'
    | 'object'
    | 'date'
    | 'array'
    | 'regexp'
    | 'function'
    | 'generatorfunction'
    | 'promise'
    | 'map'
    | 'weakmap'
    | 'set'
    | 'symbol'
    | 'int8array'
    | 'uint8array'
    | 'uint8clampedarray'
    | 'int16array'
    | 'uint16array'
    | 'int32array'
    | 'uint32array'
    | 'float32array'
    | 'float64array'

export interface Valid<T> {
    valid: true
    value: T
}

export interface Invalid<T> {
    valid: false
    errors: Errors
    message: string
}

export function invalid<T>(errors: Errors): Invalid<T> {
    return {
        valid: false,
        errors,
        get message() {
            return errorMessage(this.errors)
        }
    }
}

export function valid<T>(value: T): Valid<T> {
    return {
        valid: true,
        value
    }
}

export type ValidateResult<T> = Valid<T> | Invalid<T>
export interface ValidateFn<T> {
    (value: any): Promise<ValidateResult<T>>
}

export interface Validator<T> {
    name: string;
    reverse: (value: T) => any
    validate(value: any): Promise<ValidateResult<T>>
    or<U>(validator: Validator<U>): Validator<T|U>
    and<U>(validator: Validator<U>): Validator<T&U>
}

export function expect(kind: Kind, value: any): ValidateResult<any> {
    const k = kindOf(value)
    if (k != kind) {
        return invalid({
            type: 'type',
            name: 'expect',
            expected: kind,
            actual: k
        })
    }
    return valid(value)
}

export function typeValue<T>(v: Validator<T>): T {
    return undefined as any
}
