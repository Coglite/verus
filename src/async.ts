import { invalid, valid } from './common'
import { FluentValidator } from './fluent-validator'

export const Async = <T>(handler: (value: any) => Promise<T>) => new FluentValidator<T>('Async', async value => {
    try {
        const result = await handler(value)
        return valid(result)
    } catch (error) {
        const message = '' + error
        return invalid({
            type: 'async',
            name: 'Async',
            message
        })
    }
})