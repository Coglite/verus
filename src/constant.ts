import { FluentValidator } from './fluent-validator'
import { invalid, valid } from './common'

export const Constant = <T>(constant: T) => {
  const name = `Constant(${constant})`
  return new FluentValidator<T>(
    name,
    async (value: any) => {
      if (value !== constant) {
        return invalid({
          type: 'constant',
          name,
          expected: constant,
          actual: value,
        })
      }
      return valid(value)
    })
}