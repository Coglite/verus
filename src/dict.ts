import { FluentValidator } from './fluent-validator'
import { Validator, invalid, valid } from './common'
import * as kindOf from 'kind-of'
import { DictError } from './errors'

export const Dict = <T>(validator: Validator<T>) => {
    const name = `Dict(${validator.name})`
    return new FluentValidator<{ [key: string]: T }>(name, async value => {
        const type = kindOf(value)
        if (type !== 'object') {
            return invalid({
                type: 'type',
                name,
                expected: 'object',
                actual: type
            })
        }

        const keys = Object.keys(value)
        const values: {[key: string]: T} = {}
        const errors: DictError = {
            type: 'dict',
            name,
            fields: {}
        }
        let isValid = true

        for (const key of keys) {
            const r = await validator.validate(value[key])
            if (!r.valid) {
                isValid = false
                errors.fields[key] = r.errors
            }
            if (isValid && r.valid) {
                values[key] = r.value
            }
        }

        if (!isValid) {
            return invalid(errors)
        }
        return valid(values)
    })
}