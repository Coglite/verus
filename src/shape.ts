import { Validator, ValidateResult, ValidateFn, invalid, valid } from './common'
import { String, Number } from './type'
import { ShapeErrors } from './errors'
import * as kindOf from 'kind-of'
import { FluentValidator } from './fluent-validator'

export type ShapeFields<T> = {
    [P in keyof T]: Validator<T[P]>
}

export interface ShapeClass<T> extends Validator<T> {
    new (): T
}

export function Shape<T>(fields: ShapeFields<T>, allowExtraFields = true): ShapeClass<T> {
    class Shape {
        static async validate(value: any): Promise<ValidateResult<T>> {
            const o: Partial<T> = new this() as any
            let isValid = true
            let errors: ShapeErrors = {
                type: 'shape',
                name: 'Shape',
                shapeName: this.name,
                fields: {}
            }
            const shapeKeys = Object.keys(fields)
            for (const key of shapeKeys) {
                const r = await fields[key].validate(value[key])
                if (!r.valid) {
                    errors.fields[key] = r.errors
                    isValid = false
                } else if (isValid) {
                    o[key] = r.value
                }
            }
            const extraKeys = difference(shapeKeys, Object.keys(value))
            if (allowExtraFields) {
                for (const key of extraKeys) {
                    o[key] = value[key]
                }
            } else if (extraKeys.length > 0) {
                errors.extra = {}
                for (const key of extraKeys) {
                    errors.extra[key] = kindOf(value[key])
                }
                return invalid(errors)
            }

            return isValid ? valid(o as T) : invalid(errors)
        }

        static or<U>(other: Validator<U>): Validator<T | U> {
            const name = `${this.name} | ${other.name}`
            const v = async (value: any): Promise<ValidateResult<T | U>> => {
                const ra = await this.validate(value)
                if (ra.valid) {
                    return ra as ValidateResult<T | U>
                } else {
                    const rb = await other.validate(value)
                    if (!rb.valid) {
                        return invalid({
                            type: 'or',
                            name,
                            left: ra.errors,
                            right: rb.errors,
                        })
                    }
                    return rb
                }
            }
            return new FluentValidator<T | U>(name, v)
        }

        static and<U>(other: Validator<U>): Validator<T & U> {
            const name = `${this.name} & ${other.name}`
            const v = async (value: any): Promise<ValidateResult<T & U>> => {
                const ra = await this.validate(value)
                if (ra.valid) {
                    const rb = await other.validate(ra.value)
                    if (rb.valid) {
                        return valid(rb.value as T & U)
                    } else {
                        return invalid({
                            type: 'and',
                            name,
                            right: rb.errors
                        })
                    }
                } else {
                    return invalid({
                        type: 'and',
                        name,
                        left: ra.errors
                    })
                }
            }

            return new FluentValidator<T & U>(name, v)
        }

        static get typeValue(): T {
            throw 'Do not call typeValue at runtime'
        }

        static reverse(value: T): any {
            const o: any = {}
            for (const key of Object.keys(fields)) {
                const field = fields[key]
                o[key] = field.reverse((value as any)[key])
            }
            const diff = difference(Object.keys(fields), Object.keys(value))
            for (const key of diff) {
                o[key] = (value as any)[key]
            }
            return o
        }
    }
    return Shape as any
}

function difference(a: string[], b: string[]): string[] {
    const diff = []
    for (const vb of b) {
        if (a.indexOf(vb) === -1) {
            diff.push(vb)
        }
    }
    return diff
}
