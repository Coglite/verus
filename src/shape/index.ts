import { Validator, ValidateResult, invalid } from '../common'
import { FluentValidator } from '../fluent-validator'
import * as kindOf from 'kind-of'
import { ShapeErrors } from '../errors'

export type ShapeFields<T> = {
    [P in keyof T]: Validator<T[P]>
}

export function Shape<T>(shapeName: string, fields: ShapeFields<T>, allowExtraFields = true): FluentValidator<T> {
    const v = async (value: any): Promise<ValidateResult<T>> => {
        const o: Partial<T> = {}
        let valid = true
        let errors: ShapeErrors = {
            type: 'shape',
            name: 'Shape',
            shapeName: shapeName,
            fields: {}
        }
        const shapeKeys = Object.keys(fields)
        for (const key of shapeKeys) {
            const r = await fields[key].validate(value[key])
            if (r.result === 'invalid') {
                errors.fields[key] = r.errors
                valid = false
            } else if (valid) {
                o[key] = r.value
            }
        }
        const extraKeys = difference(shapeKeys, Object.keys(value))
        if (allowExtraFields) {
            for (const key of extraKeys) {
                o[key] = value[key]
            }
        } else if (extraKeys.length > 0) {
            errors.extra = {}
            for (const key of extraKeys) {
                errors.extra[key] = kindOf(value[key])
            }
            return invalid(errors)
        }

        return valid ? { result: 'valid', value: o as T} : invalid(errors)
    }
    return new FluentValidator<T>(`Shape(${shapeName})`, v)
}

function difference(a: string[], b: string[]): string[] {
    const diff = []
    for (const vb of b) {
        if (a.indexOf(vb) === -1) {
            diff.push(vb)
        }
    }
    return diff
}
