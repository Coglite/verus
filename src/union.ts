import { FluentValidator } from './fluent-validator'
import { invalid, valid } from './common'

export function Union<
  A extends string,
  B extends string,
  C extends string,
  D extends string,
  E extends string,
  F extends string,
  G extends string,
  H extends string,
  I extends string,
  J extends string,
  K extends string
>(
  a: A,
  b: B,
  c: C,
  d: D,
  e: E,
  f: F,
  g: G,
  h: H,
  i: I,
  j: J,
  k: K
): FluentValidator<A | B | C | D | E | F | G | H | I | J | K>
export function Union<
  A extends string,
  B extends string,
  C extends string,
  D extends string,
  E extends string,
  F extends string,
  G extends string,
  H extends string,
  I extends string,
  J extends string
>(
  a: A,
  b: B,
  c: C,
  d: D,
  e: E,
  f: F,
  g: G,
  h: H,
  i: I,
  j: J
): FluentValidator<A | B | C | D | E | F | G | H | I | J>
export function Union<
  A extends string,
  B extends string,
  C extends string,
  D extends string,
  E extends string,
  F extends string,
  G extends string,
  H extends string,
  I extends string
>(
  a: A,
  b: B,
  c: C,
  d: D,
  e: E,
  f: F,
  g: G,
  h: H,
  i: I
): FluentValidator<A | B | C | D | E | F | G | H | I>
export function Union<
  A extends string,
  B extends string,
  C extends string,
  D extends string,
  E extends string,
  F extends string,
  G extends string,
  H extends string
>(
  a: A,
  b: B,
  c: C,
  d: D,
  e: E,
  f: F,
  g: G,
  h: H
): FluentValidator<A | B | C | D | E | F | G | H>
export function Union<
  A extends string,
  B extends string,
  C extends string,
  D extends string,
  E extends string,
  F extends string,
  G extends string
>(
  a: A,
  b: B,
  c: C,
  d: D,
  e: E,
  f: F,
  g: G
): FluentValidator<A | B | C | D | E | F | G>
export function Union<
  A extends string,
  B extends string,
  C extends string,
  D extends string,
  E extends string,
  F extends string
>(a: A, b: B, c: C, d: D, e: E, f: F): FluentValidator<A | B | C | D | E | F>
export function Union<
  A extends string,
  B extends string,
  C extends string,
  D extends string,
  E extends string
>(a: A, b: B, c: C, d: D, e: E): FluentValidator<A | B | C | D | E>
export function Union<
  A extends string,
  B extends string,
  C extends string,
  D extends string
>(a: A, b: B, c: C, d: D): FluentValidator<A | B | C | D>
export function Union<A extends string, B extends string, C extends string>(
  a: A,
  b: B,
  c: C
): FluentValidator<A | B | C>
export function Union<A extends string, B extends string>(
  a: A,
  b: B
): FluentValidator<A | B>
export function Union(...values: string[]) {
  const name = `Union(${values.join('|')})`
  return new FluentValidator<any>(name, async value => {
    if (values.indexOf(value) === -1) {
      return invalid({
        type: 'union',
        name,
      })
    }
    return valid(value)
  })
}
