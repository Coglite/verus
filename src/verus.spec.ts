import {
  String,
  FluentValidator,
  Number,
  ISODate,
  ArrayOf,
  typeValue,
  errorMessage,
} from './index'
import { Shape } from './shape'
import { Invalid } from './common'
import { Union } from './union'
import { Async } from './async'
import { Dict } from './dict'
import { Constant } from './constant'

class User extends Shape({
  name: String,
}) {}

class NoExtraUser extends Shape(
  {
    name: String,
  },
  false
) {}

const IsMock = () =>
  new FluentValidator<any>(
    'Mock',
    jest.fn(value => ({ valid: true, value }))
  ) as FluentValidator<any> & { validate: typeof jest }

describe('Shape', () => {
  it('should validate simple shape', async () => {
    const result = await User.validate({ name: 'Jane Smith', dogs: ['Fluffy'] })
    expect(result).toMatchSnapshot()
  })

  it('should detect invalid simple shape', async () => {
    const result = await User.validate({ name: 234, dogs: ['Fluffy'] })
    expect(result).toMatchSnapshot()
  })

  it('should optionally not allow extra fields in value', async () => {
    let result = await NoExtraUser.validate({
      name: 'Jane Smith',
      dogs: ['Fluffy'],
    })
    expect(result).toMatchSnapshot()

    result = await NoExtraUser.validate({
      name: 'Jane Smith',
    })

    expect(result).toMatchSnapshot()

    result = await NoExtraUser.validate({
      name: 123,
      dogs: ['Snuffles'],
    })

    expect(result).toMatchSnapshot()
  })
})

describe('FluentValidator', () => {
  it('should not evaluate the right side if left side fails', async () => {
    const m = IsMock()
    const stringAndMock = String.and(m)
    const { valid } = await stringAndMock.validate(231)
    expect(valid).toBe(false)
    expect(m.validate).toHaveBeenCalledTimes(0)
  })

  it('should evaluate both conditions if left side is valid', async () => {
    const m = IsMock()
    const stringAndMock = String.and(m)
    const { valid } = await stringAndMock.validate('foobar')
    expect(valid).toBe(true)
    expect(m.validate).toHaveBeenCalledTimes(1)
  })
})

describe('ISODate', () => {
  it('should not allow invalid dates', async () => {
    const r = await ISODate.validate('qwerqwerqwer')
    expect(r).toMatchSnapshot()
  })

  it('should parse ISO 8601 dates', async () => {
    const r = await ISODate.validate('2007-04-05T14:30-02:00')
    expect(r).toMatchSnapshot()
  })

  it('should not allow non-string values', async () => {
    const r = await ISODate.validate(324)
    expect(r).toMatchSnapshot()
  })
})

describe('ArrayOf', () => {
  it('should validate an array of strings', async () => {
    const r = await ArrayOf(String).validate(['foo', 'bar', 'baz'])
    expect(r).toMatchSnapshot()
  })

  it('should transform values to a new array', async () => {
    const r = await ArrayOf(ISODate).validate([
      '2007-04-05T14:30-02:00',
      '2008-04-05T14:30-02:00',
      '2009-04-05T14:30-02:00',
      '2010-11-05T14:30-02:00',
    ])
    expect(r).toMatchSnapshot()
  })

  it('should fail if some values are invalid', async () => {
    const r = await ArrayOf(ISODate).validate([
      '2009-04-05T14:30-02:00',
      '2015-06lasdkjflöakjsdfölkja',
      '2010-11-05T14:30-02:00',
      '2015-06lasdka',
    ])
    expect(r).toMatchSnapshot()
  })

  it('should fail on non-array value', async () => {
    const r = await ArrayOf(ISODate).validate(42)
    expect(r).toMatchSnapshot()
  })
})

describe('String', () => {
  it('should not validate if value is not string', async () => {
    const r = await String.validate(2313)
    expect(r).toMatchSnapshot()
  })
  it('should validate if length > minLength', async () => {
    const r = await String.minLength(4).validate('qwerty')
    expect(r).toMatchSnapshot()
  })

  it('should validate if length < maxLength', async () => {
    const r = await String.maxLength(4).validate('qwe')
    expect(r).toMatchSnapshot()
  })

  it('should fail if length < minLength', async () => {
    const r = await String.minLength(4).validate('qwe')
    expect(r).toMatchSnapshot()
  })

  it('should fail if length > maxLength', async () => {
    const r = await String.maxLength(3).validate('qwer')
    expect(r).toMatchSnapshot()
  })
})

describe('Union', () => {
  it('should match valid value', async () => {
    const r = await Union('a', 'b').validate('b')
    expect(r).toMatchSnapshot()
  })

  it('should not match invalid value', async () => {
    const r = await Union('a', 'b', 'c').validate('d')
    expect(r).toMatchSnapshot()
  })
})

describe('Async', () => {
  it('should succeed on valid value', async () => {
    const asyncBool = Async(async value => true)
    const r = await asyncBool.validate('')
    expect(r).toMatchSnapshot()
  })

  it('should fail on invalid value', async () => {
    const asyncBool = Async(async value => {
      throw 'is invalid!'
    })
    const r = await asyncBool.validate('')
    expect(r).toMatchSnapshot()
  })
})

class Kid extends Shape({
  name: String,
  birthDate: ISODate,
}) {}

class Relative extends Shape({
  name: String,
  birthDate: ISODate,
  kids: ArrayOf(Kid),
}) {}

describe('Validator.reverse', () => {
  it('should produce the same result before the validation', async () => {
    const relative = {
      name: 'Joan',
      birthDate: '1982-08-24T17:26:59.226Z',
      kids: [
        {
          name: 'Jack',
          birthDate: '2007-08-24T17:26:59.226Z',
        },
        {
          name: 'Jill',
          birthDate: '2007-08-24T17:29:59.226Z',
        },
      ],
      extra: 'extra data here',
    }
    const result = await Relative.validate(relative)
    expect(result.valid).toBe(true)

    if (!result.valid) throw ''
    const reversed = Relative.reverse(result.value)
    expect(reversed).toEqual(relative)
  })
})

describe('Dict', () => {
  it('should succeed on valid value', async () => {
    const stringDict = Dict(Number)
    const r = await stringDict.validate({
      hello: 324,
      world: 123,
    })
    expect(r).toMatchSnapshot()
  })

  it('should fail on invalid values', async () => {
    const stringDict = Dict(Number)
    const r = await stringDict.validate({
      hello: 324,
      world: 'qwer',
      foo: 435,
      bar: new Date(),
    })
    expect(r).toMatchSnapshot()
  })
})

describe('Constant', () => {
  it('should succeed on valid value', async () => {
    const constant = Constant<'asdf'>('asdf')
    const r = await constant.validate('asdf')
    expect(r).toMatchSnapshot()
  })

  it('should fail on invalid values', async () => {
    const constant = Constant<'asdf'>('asdf')
    const r = await constant.validate('qwer')
    expect(r).toMatchSnapshot()
  })
})

class Product extends Shape({
  name: String,
  price: Number,
}) {}

class Subscription extends Shape({
  name: String,
  startDate: ISODate,
  endDate: ISODate,
  monthlyPrice: Number,
}) {}

class Gift extends Shape({
  name: String,
  recipient: String,
  price: Number,
}) {}

class Cart extends Shape({
  created: ISODate,
  items: ArrayOf(Product.or(Subscription).or(Gift)),
}) {}

describe('Error messages', () => {
  it('should produce 3 lines of error messages', async () => {
    const r = await Cart.validate({
      created: '2009-04-05garbage',
      items: [
        { name: 'Candybar', price: '0.9' },
        { name: 'Bread', price: 3.4 },
        {
          name: 'DSL',
          date: '2009-04-05T14:30-02:00',
          endDate: '2010-02-05T14:30-02:00',
          monthlyPrice: 33.4,
        },
      ],
    })
    expect(r.valid).toBe(false)
    expect((r as Invalid<any>).message).toMatchSnapshot()
  })
})
