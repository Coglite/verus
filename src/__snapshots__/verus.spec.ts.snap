// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArrayOf should fail if some values are invalid 1`] = `
Object {
  "errors": Object {
    "invalidIndices": Array [
      Array [
        1,
        Object {
          "expectedFormat": "ISO 8601 date",
          "name": "ISODate",
          "type": "parse",
        },
      ],
      Array [
        3,
        Object {
          "expectedFormat": "ISO 8601 date",
          "name": "ISODate",
          "type": "parse",
        },
      ],
    ],
    "name": "ArrayOf<ISODate>",
    "type": "array",
  },
  "message": "[1] did not match 'ISO 8601 date'-format
[3] did not match 'ISO 8601 date'-format
",
  "valid": false,
}
`;

exports[`ArrayOf should fail on non-array value 1`] = `
Object {
  "errors": Object {
    "actual": "number",
    "expected": "array",
    "name": "expect",
    "type": "type",
  },
  "message": "value is the wrong type, expected array, got number
",
  "valid": false,
}
`;

exports[`ArrayOf should transform values to a new array 1`] = `
Object {
  "valid": true,
  "value": Array [
    2007-04-05T16:30:00.000Z,
    2008-04-05T16:30:00.000Z,
    2009-04-05T16:30:00.000Z,
    2010-11-05T16:30:00.000Z,
  ],
}
`;

exports[`ArrayOf should validate an array of strings 1`] = `
Object {
  "valid": true,
  "value": Array [
    "foo",
    "bar",
    "baz",
  ],
}
`;

exports[`Async should fail on invalid value 1`] = `
Object {
  "errors": Object {
    "message": "is invalid!",
    "name": "Async",
    "type": "async",
  },
  "message": "value failed async validation: \\"is invalid!\\"
",
  "valid": false,
}
`;

exports[`Async should succeed on valid value 1`] = `
Object {
  "valid": true,
  "value": true,
}
`;

exports[`Dict should fail on invalid values 1`] = `
Object {
  "errors": Object {
    "fields": Object {
      "bar": Object {
        "actual": "date",
        "expected": "number",
        "name": "Number",
        "type": "type",
      },
      "world": Object {
        "actual": "string",
        "expected": "number",
        "name": "Number",
        "type": "type",
      },
    },
    "name": "Dict(Number)",
    "type": "dict",
  },
  "message": "Dict(Number)['world'] is the wrong type, expected number, got string
Dict(Number)['bar'] is the wrong type, expected number, got date
",
  "valid": false,
}
`;

exports[`Dict should succeed on valid value 1`] = `
Object {
  "valid": true,
  "value": Object {
    "hello": 324,
    "world": 123,
  },
}
`;

exports[`Error messages should produce 3 lines of error messages 1`] = `
"Cart.created did not match 'ISO 8601 date'-format
Cart.items[0] failed all validators: Product | Subscription | Gift
Cart.items[2] failed all validators: Product | Subscription | Gift
"
`;

exports[`ISODate should not allow invalid dates 1`] = `
Object {
  "errors": Object {
    "expectedFormat": "ISO 8601 date",
    "name": "ISODate",
    "type": "parse",
  },
  "message": "value did not match 'ISO 8601 date'-format
",
  "valid": false,
}
`;

exports[`ISODate should not allow non-string values 1`] = `
Object {
  "errors": Object {
    "actual": "number",
    "expected": "string",
    "name": "expect",
    "type": "type",
  },
  "message": "value is the wrong type, expected string, got number
",
  "valid": false,
}
`;

exports[`ISODate should parse ISO 8601 dates 1`] = `
Object {
  "valid": true,
  "value": 2007-04-05T16:30:00.000Z,
}
`;

exports[`Shape should detect invalid simple shape 1`] = `
Object {
  "errors": Object {
    "fields": Object {
      "name": Object {
        "actual": "number",
        "expected": "string",
        "name": "String",
        "type": "type",
      },
    },
    "name": "Shape",
    "shapeName": "User",
    "type": "shape",
  },
  "message": "User.name is the wrong type, expected string, got number
",
  "valid": false,
}
`;

exports[`Shape should optionally not allow extra fields in value 1`] = `
Object {
  "errors": Object {
    "extra": Object {
      "dogs": "array",
    },
    "fields": Object {},
    "name": "Shape",
    "shapeName": "NoExtraUser",
    "type": "shape",
  },
  "message": "NoExtraUser.dogs is an unexpected field with type array
",
  "valid": false,
}
`;

exports[`Shape should optionally not allow extra fields in value 2`] = `
Object {
  "valid": true,
  "value": NoExtraUser {
    "name": "Jane Smith",
  },
}
`;

exports[`Shape should optionally not allow extra fields in value 3`] = `
Object {
  "errors": Object {
    "extra": Object {
      "dogs": "array",
    },
    "fields": Object {
      "name": Object {
        "actual": "number",
        "expected": "string",
        "name": "String",
        "type": "type",
      },
    },
    "name": "Shape",
    "shapeName": "NoExtraUser",
    "type": "shape",
  },
  "message": "NoExtraUser.name is the wrong type, expected string, got number
NoExtraUser.dogs is an unexpected field with type array
",
  "valid": false,
}
`;

exports[`Shape should validate simple shape 1`] = `
Object {
  "valid": true,
  "value": User {
    "dogs": Array [
      "Fluffy",
    ],
    "name": "Jane Smith",
  },
}
`;

exports[`String should fail if length < minLength 1`] = `
Object {
  "errors": Object {
    "length": 3,
    "minLength": 4,
    "name": "String",
    "type": "length",
  },
  "message": "value is too short, min length is 4, was 3
",
  "valid": false,
}
`;

exports[`String should fail if length > maxLength 1`] = `
Object {
  "errors": Object {
    "length": 4,
    "maxLength": 3,
    "name": "String",
    "type": "length",
  },
  "message": "value is too long, max length is 3, was 4
",
  "valid": false,
}
`;

exports[`String should not validate if value is not string 1`] = `
Object {
  "errors": Object {
    "actual": "number",
    "expected": "string",
    "name": "String",
    "type": "type",
  },
  "message": "value is the wrong type, expected string, got number
",
  "valid": false,
}
`;

exports[`String should validate if length < maxLength 1`] = `
Object {
  "valid": true,
  "value": "qwe",
}
`;

exports[`String should validate if length > minLength 1`] = `
Object {
  "valid": true,
  "value": "qwerty",
}
`;

exports[`Union should match valid value 1`] = `
Object {
  "valid": true,
  "value": "b",
}
`;

exports[`Union should not match invalid value 1`] = `
Object {
  "errors": Object {
    "name": "Union(a|b|c)",
    "type": "union",
  },
  "message": "value did not match union Union(a|b|c)
",
  "valid": false,
}
`;
